File: 1036-RottingOranges/1036-RottingOranges.java
// Last updated: 28/06/2026, 01:02:18
class Solution {
    public int orangesRotting(int[][] grid) {
        int rows=grid.length,cols=grid[0].length;
        int fresh=0, minutes=0;
        Queue<int[]> q =new LinkedList<>();
        for(int r=0;r<rows;r++){
            for(int c=0;c<cols;c++){
                if(grid[r][c]==2)
                      q.offer(new int[] {r,c});
                if(grid[r][c]==1) 
                      fresh++;
                
            }
        }
        int[][] directions={{-1,0},{1,0},{0,-1},{0,1}};
        while(!q.isEmpty() && fresh>0){
            minutes++;
            int n=q.size();
            for(int i=0;i<n;i++){
            int[] cell =q.poll();
            int row=cell[0];
            int col=cell[1];
            for(int[] d:directions){
                int newr=row+d[0];
                int newc=col+d[1];
                if(newr>=0 && newr<rows && newc>=0 && newc<cols && grid[newr][newc]==1){
                    grid[newr][newc]=2;
                    fresh--;
                    q.offer(new int[] {newr,newc});
                }
            }
            
            }
          

        }
        return fresh==0?minutes:-1;
    }
}

File: 1036-RottingOranges/1036-RottingOranges.cpp
// Last updated: 28/06/2026, 01:02:21
struct node{
    int x,y,time;
    node(int _x,int _y,int _time)
    {
        x=_x;
        y=_y;
        time=_time;
        
    }
};

class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        queue<node>q;
        int countoranges=0;
        // push the rotten oranges
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(grid[i][j]==2)
                {
                    q.push(node(i,j,0));
                }
                if(grid[i][j]!=0)
                    countoranges++;
            }
        }
        int minimumtime=0;
        // how many times oranges are rottening
        int count =0;
        while(!q.empty())
        {
            int x=q.front().x;
            int y=q.front().y;
            int time=q.front().time;
            q.pop();
            count++;
            minimumtime=max( minimumtime,time);
            // go four directions 
            int dx[]={-1,0,+1,0};
            int dy[]={0,+1,0,-1};
            for(int ind=0;ind<4;ind++)
            {
                int newx=x+dx[ind];
                int newy=y+dy[ind];
                // boundaries check
                if(newx>=0 && newx<n && newy>=0 && newy<m && grid[newx][newy]==1)
                    // now i can rotten oranges becoz satisfy above cond.
                {
                    grid[newx][newy]=2; // make fresh oranges to rotten oranges
                    q.push(node(newx,newy,time+1));
                }
            }
            
        }
        if(count==countoranges)
            return  minimumtime;
    return -1;
    }
};